backends/blocked/ceed-blocked-operator.c

```diff
@@ -41,29 +41,54 @@ static int CeedOperatorSetupFields_Blocked(CeedQFunction qf, CeedOperator op, bo
-      } else {
-        const CeedInt *offsets = NULL;
-        CeedCallBackend(CeedElemRestrictionGetOffsets(r, CEED_MEM_HOST, &offsets));
-        CeedCallBackend(CeedElemRestrictionGetCompStride(r, &comp_stride));
-        CeedCallBackend(CeedElemRestrictionCreateBlocked(ceed, num_elem, elem_size, blk_size, num_comp, comp_stride, l_size, CEED_MEM_HOST,
-                                                         CEED_COPY_VALUES, offsets, &blk_restr[i + start_e]));
-        CeedCallBackend(CeedElemRestrictionRestoreOffsets(r, &offsets));
+      CeedCallBackend(CeedElemRestrictionGetCompStride(r, &comp_stride));
+
+      CeedRestrictionType rstr_type;
+      CeedCallBackend(CeedElemRestrictionGetType(r, &rstr_type));
+      switch (rstr_type) {
+        case CEED_RESTRICTION_STANDARD: {
+          const CeedInt *offsets = NULL;
+          CeedCallBackend(CeedElemRestrictionGetOffsets(r, CEED_MEM_HOST, &offsets));
+          CeedCallBackend(CeedElemRestrictionCreateBlocked(ceed_rstr, num_elem, elem_size, blk_size, num_comp, comp_stride, l_size, CEED_MEM_HOST,
+                                                           CEED_COPY_VALUES, offsets, &blk_restr[i + start_e]));
+          CeedCallBackend(CeedElemRestrictionRestoreOffsets(r, &offsets));
+        } break;
+        case CEED_RESTRICTION_ORIENTED: {
+          const CeedInt *offsets = NULL;
+          const bool    *orients = NULL;
+          CeedCallBackend(CeedElemRestrictionGetOffsets(r, CEED_MEM_HOST, &offsets));
+          CeedCallBackend(CeedElemRestrictionGetOrientations(r, CEED_MEM_HOST, &orients));
+          CeedCallBackend(CeedElemRestrictionCreateBlockedOriented(ceed_rstr, num_elem, elem_size, blk_size, num_comp, comp_stride, l_size,
+                                                                   CEED_MEM_HOST, CEED_COPY_VALUES, offsets, orients, &blk_restr[i + start_e]));
+          CeedCallBackend(CeedElemRestrictionRestoreOffsets(r, &offsets));
+          CeedCallBackend(CeedElemRestrictionRestoreOrientations(r, &orients));
+        } break;
+        case CEED_RESTRICTION_CURL_ORIENTED: {
+          const CeedInt  *offsets      = NULL;
+          const CeedInt8 *curl_orients = NULL;
+          CeedCallBackend(CeedElemRestrictionGetOffsets(r, CEED_MEM_HOST, &offsets));
+          CeedCallBackend(CeedElemRestrictionGetCurlOrientations(r, CEED_MEM_HOST, &curl_orients));
+          CeedCallBackend(CeedElemRestrictionCreateBlockedCurlOriented(ceed_rstr, num_elem, elem_size, blk_size, num_comp, comp_stride, l_size,
+                                                                       CEED_MEM_HOST, CEED_COPY_VALUES, offsets, curl_orients,
+                                                                       &blk_restr[i + start_e]));
+          CeedCallBackend(CeedElemRestrictionRestoreOffsets(r, &offsets));
+          CeedCallBackend(CeedElemRestrictionRestoreCurlOrientations(r, &curl_orients));
+        } break;
```

Comment:
I think I see why these aren't being tested. `t571` creates a `CeedOperator` which as I understand it, should dispatch to `CeedOperatorLinearAssembleQFunctionCore_Opt` and `CeedOperatorLinearAssembleQFunctionCore_Blocked` when we call `CeedOperatorLinearAssemble` and thus cover these code blocks. It looks like it isn't though, and every call in the test falls back to `CeedOperatorLinearAssembleQFunctionCore_Ref` in the `ref/serial` backend. I think this is because in `CeedSingleOperatorAssemble` we don't have an `op->LinearAssembleSingle` entry and thus we fall back to `ref/serial` before we actually need to. Is this intended/is there a way to improve this to make use of the `CeedOperatorLinearAssembleQFunctionCore` specializations for the `opt`/`blocked` backends?

Reply:
Yeah, that's by design. We only have backend specific variants of `CeedOperatorLinearAssembleQFunction`, not of the full operator assembly. It hasn't bottlenecked us to not have some special version of assembly for the opt operator evaluation.

So probably a separate test that directly calls `CeedOperatorApply` will fix this?

Reply:
Sorry I meant "resolve the issue" as "resolve the CPU issue" regarding prematurely falling back for `opt` and `ref/blocked`. But yes I see the problem with that simple approach now. Maybe I can think about this a bit more about it of how to make use of the `CeedOperatorLinearAssembleQFunctionCore_Opt/Blocked` here still.

Reply:
`CeedOperatorLinearAssembleQFunctionCore_Opt/Blocked` is being used. The missing lines of coverage are in the `CeedOperatorApply_Opt/Blocked` related code.

Reply:
If you add print statements to `CeedOperatorLinearAssembleQFunctionCore_Opt/Blocked`, you can see they are not being used by `t571`. This is because these backends don't have a `LinearAssembleSingle` specialization so we fall back to `ref/serial` before we get the chance to call `LinearAssembleQFunction`/`LinearAssembleQFunctionUpdate`, which they do have specializations for. If we were in fact calling the specialized `CeedOperatorLinearAssembleQFunctionCore` methods, then we would get calls to `CeedOperatorSetup_Opt/Blocked` and the test coverage would be fixed.

Reply:
👍 Agreed, done in 726c2de so that should hopefully squash the remaining coverage issues.

Reply:
Nice, I will rebase #1265 and double check with you once that is done before merging.